### ISSUE TITLE
Unify Local & Cloud authentication with a common interface

### ISSUE THREAD

**Issue by b5** [qri-io/qri#1806]
We need to harmonize the story of authentication across qri.cloud & local qri instances. The most practical thing to focus on is getting responses for two authentication routes (`/token` & `/refresh`) to be backed by a common interface with separate local & cloud implementations.

on the cloud side this interface should provide a full-fledged oauth implementation, and the local side is basically "just pick from the key pairs you have".

There needs to be come alignment between cloud & core about _how_ they use oauth. For this to work any frontend/external code needs to be able to disambiguate what type of authentication mechanism a qri node expects.

While we're here, it'd be great to also eliminate the two different flows for authentication (one for RPC, and another for API). 

The authentication interface should be declared in the auth package, and should be providable via a `lib.NewInstance` option. Ideally all auth is delegated to this interface.

**Comment by Arqu** [comment]
Drawing some basic lines around the common interface
```
package auth

import (
	...
)

type AuthProvider interface {
	// Token handles the auth token flow
	Token(ctx context.Context, req *TokenRequest) (*TokenResponse, error)
	// ParseTokenRequest extracts the TokenRequest from the incoming http request
	ParseTokenRequest(r *http.Request) (*TokenRequest, error)
	// TokenMiddleware parses any "authorization" header containing a Bearer
	// token & adds it to the request context
	TokenMiddleware(next http.Handler) http.Handler
}

// TokenRequest is a wrapper for incoming token requests
type TokenRequest struct {
	GrantType    GrantType
	Code         string
	Username     string
	Password     string
	RefreshToken string
	RedirectURI  string
}

// TokenResponse wraps the token response object
type TokenResponse struct {
	AccessToken  string `json:"access_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int64  `json:"expires_in"`
	RefreshToken string `json:"refresh_token,omitempty"`
}
```

The interface is quite simple in that the intersection between local & cloud and to make it work we only need to support getting a token (along with refreshing it). The middleware is only forced on here because down the line ideally most cloud request would flow through the core implementation in some shape or form, so it's important to be able to inject a custom middleware for each which process the token slightly differently. Standalone `qri` is meant to set the `scope.ActiveProfile` while on cloud it's intentionally kept separate as we have a global shared state via the `datastore` and will not be ready to move to scope in the near term.

To round out the update it would be great to have most of the "guts" move into qri core which would mean qri absorbs some of the generic cloud features around token claims. In turn cloud would drop most of it's internal definitions in favor of those.

The middleware will in both cases set the token to the request context, which on cloud is used throughout the lifetime of a request, while on core this responsibility is on `scope` and currently the only missing piece seems to be setting the correct `activeProfile` based on the token (if provided) on scope creation. (Along with some double checking to make sure we really do only use the profile from scope internally).

**Comment by Arqu** [comment]
@dustmop @b5 @ramfox ^ 
Comments are appreciated, I'll get started on this in the morning.

**Comment by b5** [comment]
Looks good! A few thoughts & questions:

`TokenResponse` json field tags should be lowerCamelCase.

with regard to `ParseTokenRequest`, how will the different implementations differ in the way they hydrate `TokenRequest`? Would it be possible to factor that into a function defined in the core `auth` package? Feels like it could be uniform, and the `AuthProvider` implementations could reject token requests of a type they can't work with. Might require adding fields to the `TokenRequest` param struct, but I think that's fine. We could also define a sentinel `ErrCannotHandleTokenType` err that implementations can respond with.

> The middleware will in both cases set the token to the request context

Similar thinking for the `Middleware` method. I'm wondering how different each implementation really needs to be, and if we can't get away with a `Middleware` factory function that parses tokens from `*http.Request`'s, and works with the existing `auth.AddToContext` function to inject a token into the context. this could happen before scope construction, and as you said, cloud just works with raw contexts

If those are possible, we can get `AuthProvider` down to a single method & think about renaming it to `TokenProvider`

**Comment by Arqu** [comment]
> Looks good! A few thoughts & questions:
> 
> `TokenResponse` json field tags should be lowerCamelCase.
> 
Done!
> with regard to `ParseTokenRequest`, how will the different implementations differ in the way they hydrate `TokenRequest`? Would it be possible to factor that into a function defined in the core `auth` package? Feels like it could be uniform, and the `AuthProvider` implementations could reject token requests of a type they can't work with. Might require adding fields to the `TokenRequest` param struct, but I think that's fine. We could also define a sentinel `ErrCannotHandleTokenType` err that implementations can respond with.
> 
So, about 50% of the validation logic lives in the `ParseTokenRequest` function. Initially the target was to completely ignore OAuth implementation flows we we don't intend to support on a specific platform. However, it's not too much extra baggage and I'd be fine with moving the responsibility into the `Token()` func and have it shared across implementations.
> > The middleware will in both cases set the token to the request context
> 
> Similar thinking for the `Middleware` method. I'm wondering how different each implementation really needs to be, and if we can't get away with a `Middleware` factory function that parses tokens from `*http.Request`'s, and works with the existing `auth.AddToContext` function to inject a token into the context. this could happen before scope construction, and as you said, cloud just works with raw contexts
> 
Kind of, while the above is true, and functionally very close, we intentionally want to use a different context key for now to carry the token. The reason being, on cloud while *most* flows are cloud specific and directly access the request context (and are thus isolated from the `scope.ActiveProfile()`) we do have some flows that go through handlers provided directly by core. As such we do want to keep the active profile to the one of cloud (we do have a live qri instance in there) to keep behavior on cloud consistent with what we have now until we have a deeper integration with core and a more supported multi-tenant environment.
> If those are possible, we can get `AuthProvider` down to a single method & think about renaming it to `TokenProvider`

Happy with the rename in general, `AuthProvider` was just to get started with something. 

In short, can do both, not too keen on the middleware piece purely out of clearer control, however we could hack around it.

**Comment by dustmop** [comment]
This sounds fine to me. I still feel like I'm not super familiar with the authentication code paths, so I don't have too much to add. What I do understand about it seems to align with what's being described here.
